app/com/lynxanalytics/biggraph/frontend_operations/ImportOperations.scala

```diff
@@ -151,27 +145,113 @@ class ImportOperations(env: SparkFreeEnvironment) extends OperationRegistry {
     }
   })
 
-  register("Import Neo4j")(new ImportOperation(_) {
+  register("Import from Neo4j", List(), List("graph"))(new SimpleOperation(_) with Importer {
+    import play.api.libs.json
+    import com.lynxanalytics.biggraph.controllers._
+
     params ++= List(
-      Param("node_label", "Node label"),
-      Param("relationship_type", "Relationship type"),
-      NonNegInt("num_partitions", "Number of partitions", default = 0),
-      Choice("infer", "Infer types", options = FEOption.noyes),
-      Code("sql", "SQL", language = "sql"),
-      Param("imported_columns", "Columns to import"),
-      Param("limit", "Limit"),
+      Param("url", "Neo4j connection", defaultValue = "bolt://localhost:7687"),
+      Param("username", "Neo4j username", defaultValue = "neo4j"),
+      Param("password", "Neo4j password", defaultValue = "neo4j"),
+      Param("query", "Cypher query", defaultValue = "MATCH (src)-[rel]->(dst) RETURN src, dst, rel"),
       ImportedTableParam("imported_table"),
       new DummyParam("last_settings", ""))
 
-    def getRawDataFrame(context: spark.sql.SQLContext) = {
-      Neo4jUtil.read(
-        context,
-        params("node_label"),
-        params("relationship_type"),
-        splitParam("imported_columns").toSet,
-        params("infer") == "yes",
-        params("limit"),
-        params("num_partitions").toInt)
+    override def getOutputs(): Map[BoxOutput, BoxOutputState] = {
+      params.validate()
+      assert(params("imported_table").nonEmpty, "You have to import the data first.")
+      assert(!areSettingsStale, areSettingsStaleReplyMessage)
+      import CommonProjectState._
+      val state = json.Json.parse(params("imported_table")).as[CommonProjectState]
+      val project = new RootProjectEditor(state)
+      Map(context.box.output(context.meta.outputs(0)) -> BoxOutputState.from(project))
+    }
+
+    override def runImport(env: com.lynxanalytics.biggraph.BigGraphEnvironment): ImportResult = {
+      import org.apache.spark.sql.functions._
+      import org.apache.spark.sql.types
+      import com.lynxanalytics.biggraph.graph_util.Scripting._
+      val spark = SQLController.defaultContext()(env.sparkDomain)
+      val project = new RootProjectEditor(CommonProjectState.emptyState)
+      val df = spark.read.format("org.neo4j.spark.DataSource")
+        .option("authentication.type", "basic")
+        .option("authentication.basic.username", params("username"))
+        .option("authentication.basic.password", params("password"))
+        .option("url", params("url"))
+        .option("schema.strategy", "string")
+        .option("query", params("query"))
+        .load
+        .cache
+
+      def getBy(columns: Seq[String], key: String, id: String = null) = {
```

Comment:
Huh, this is pretty involved... I'm kinda hoping this might go away with the two query setup, but if not, then it is worth a few lines of explanation.

Reply:
> Huh, this is pretty involved... I'm kinda hoping this might go away with the two query setup, but if not, then it is worth a few lines of explanation.

It's independent of whether we do one or two queries. But it's gone now.

The Neo4j connector was failing on schema detection if I had more than one kind of label in the result. So I thought I had to split by label before parsing the JSON. But no. It works fine if I don't split. It's also much faster. 😓